File: assistant/ai-assistant-api-old-python/callback.py
"""Callback handlers used in the app."""
from typing import Any, Dict, List

from langchain.callbacks.base import BaseCallbackHandler
from langchain.schema import AgentAction, AgentFinish, LLMResult
from typing import Any, Dict, List, Union

from schemas import ChatResponse


class StreamingLLMCallbackHandler(BaseCallbackHandler):
    """Callback handler for streaming LLM responses."""
    chain_starts: int = 0
    chain_ends: int = 0
    llm_starts: int = 0
    llm_ends: int = 0
    llm_streams: int = 0
    tool_starts: int = 0
    tool_ends: int = 0
    agent_ends: int = 0
    starts: int = 0
    ends: int = 0
    errors: int = 0

    def __init__(self, websocket):
        self.websocket = websocket

    async def on_llm_new_token(self, token: str, **kwargs: Any) -> None:
        resp = ChatResponse(sender="bot", message=token, type="stream")
        await self.websocket.send_json(resp.dict())

    def on_llm_start(
        self, serialized: Dict[str, Any], prompts: List[str], **kwargs: Any
    ) -> None:
        """Run when LLM starts running."""
        self.llm_starts += 1
        self.starts += 1

    def on_llm_end(self, response: LLMResult, **kwargs: Any) -> None:
        """Run when LLM ends running."""
        self.llm_ends += 1
        self.ends += 1

    def on_llm_error(
        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
    ) -> None:
        """Run when LLM errors."""
        self.errors += 1

    def on_chain_start(
        self, serialized: Dict[str, Any], inputs: Dict[str, Any], **kwargs: Any
    ) -> None:
        """Run when chain starts running."""
        self.chain_starts += 1
        self.starts += 1

    def on_chain_end(self, outputs: Dict[str, Any], **kwargs: Any) -> None:
        """Run when chain ends running."""
        self.chain_ends += 1
        self.ends += 1

    def on_chain_error(
        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
    ) -> None:
        """Run when chain errors."""
        self.errors += 1

    def on_tool_start(
        self, serialized: Dict[str, Any], input_str: str, **kwargs: Any
    ) -> None:
        """Run when tool starts running."""
        self.tool_starts += 1
        self.starts += 1

    def on_tool_end(self, output: str, **kwargs: Any) -> None:
        """Run when tool ends running."""
        self.tool_ends += 1
        self.ends += 1

    def on_tool_error(
        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
    ) -> None:
        """Run when tool errors."""
        self.errors += 1

    def on_text(self, text: str, **kwargs: Any) -> None:
        """Run when agent is ending."""
        self.text += 1

    def on_agent_finish(self, finish: AgentFinish, **kwargs: Any) -> None:
        """Run when agent ends running."""
        self.agent_ends += 1
        self.ends += 1

    def on_agent_action(self, action: AgentAction, **kwargs: Any) -> Any:
        """Run on agent action."""
        self.tool_starts += 1
        self.starts += 1


class FollowupQuestionGenCallbackHandler(BaseCallbackHandler):
    """Callback handler for question generation."""
    chain_starts: int = 0
    chain_ends: int = 0
    llm_starts: int = 0
    llm_ends: int = 0
    llm_streams: int = 0
    tool_starts: int = 0
    tool_ends: int = 0
    agent_ends: int = 0
    starts: int = 0
    ends: int = 0
    errors: int = 0

    def __init__(self, websocket):
        self.websocket = websocket

    async def on_llm_new_token(self, token: str, **kwargs: Any) -> None:
        resp = ChatResponse(sender="bot", message=token, type="stream_followup")
        print(token)
        await self.websocket.send_json(resp.dict())

    def on_llm_start(
        self, serialized: Dict[str, Any], prompts: List[str], **kwargs: Any
    ) -> None:
        """Run when LLM starts running."""
        self.llm_starts += 1
        self.starts += 1

    def on_llm_end(self, response: LLMResult, **kwargs: Any) -> None:
        """Run when LLM ends running."""
        self.llm_ends += 1
        self.ends += 1

    def on_llm_error(
        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
    ) -> None:
        """Run when LLM errors."""
        self.errors += 1

    def on_chain_start(
        self, serialized: Dict[str, Any], inputs: Dict[str, Any], **kwargs: Any
    ) -> None:
        """Run when chain starts running."""
        self.chain_starts += 1
        self.starts += 1

    def on_chain_end(self, outputs: Dict[str, Any], **kwargs: Any) -> None:
        """Run when chain ends running."""
        self.chain_ends += 1
        self.ends += 1

    def on_chain_error(
        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
    ) -> None:
        """Run when chain errors."""
        self.errors += 1

    def on_tool_start(
        self, serialized: Dict[str, Any], input_str: str, **kwargs: Any
    ) -> None:
        """Run when tool starts running."""
        self.tool_starts += 1
        self.starts += 1

    def on_tool_end(self, output: str, **kwargs: Any) -> None:
        """Run when tool ends running."""
        self.tool_ends += 1
        self.ends += 1

    def on_tool_error(
        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
    ) -> None:
        """Run when tool errors."""
        self.errors += 1

    def on_text(self, text: str, **kwargs: Any) -> None:
        """Run when agent is ending."""
        self.text += 1

    def on_agent_finish(self, finish: AgentFinish, **kwargs: Any) -> None:
        """Run when agent ends running."""
        self.agent_ends += 1
        self.ends += 1

    def on_agent_action(self, action: AgentAction, **kwargs: Any) -> Any:
        """Run on agent action."""
        self.tool_starts += 1
        self.starts += 1

File: assistant/ai-assistant-api-unused/callback.ts
import { BaseCallbackHandler } from './base.js';
import { AgentAction, AgentFinish, LLMResult } from '../schema';
import { ChatResponse } from '../schemas';

export class StreamingLLMCallbackHandler extends BaseCallbackHandler {
  chain_starts: number = 0;
  chain_ends: number = 0;
  llm_starts: number = 0;
  llm_ends: number = 0;
  llm_streams: number = 0;
  tool_starts: number = 0;
  tool_ends: number = 0;
  agent_ends: number = 0;
  starts: number = 0;
  ends: number = 0;
  errors: number = 0;

  constructor(private websocket: any) {
    super();
  }

  async on_llm_new_token(token: string, kwargs: any): Promise<void> {
    const resp: ChatResponse = {
      sender: 'bot',
      message: token,
      type: 'stream'
    };
    await this.websocket.send_json(resp);
  }

  on_llm_start(serialized: any, prompts: string[], kwargs: any): void {
    this.llm_starts += 1;
    this.starts += 1;
  }

  on_llm_end(response: LLMResult, kwargs: any): void {
    this.llm_ends += 1;
    this.ends += 1;
  }

  on_llm_error(error: Exception | KeyboardInterrupt, kwargs: any): void {
    this.errors += 1;
  }

  on_chain_start(serialized: any, inputs: any, kwargs: any): void {
    this.chain_starts += 1;
    this.starts += 1;
  }

  on_chain_end(outputs: any, kwargs: any): void {
    this.chain_ends += 1;
    this.ends += 1;
  }

  on_chain_error(error: Exception | KeyboardInterrupt, kwargs: any): void {
    this.errors += 1;
  }

  on_tool_start(serialized: any, input_str: string, kwargs: any): void {
    this.tool_starts += 1;
    this.starts += 1;
  }

  on_tool_end(output: string, kwargs: any): void {
    this.tool_ends += 1;
    this.ends += 1;
  }

  on_tool_error(error: Exception | KeyboardInterrupt, kwargs: any): void {
    this.errors += 1;
  }

  on_text(text: string, kwargs: any): void {
    this.text += 1;
  }

  on_agent_finish(finish: AgentFinish, kwargs: any): void {
    this.agent_ends += 1;
    this.ends += 1;
  }

  on_agent_action(action: AgentAction, kwargs: any): any {
    this.tool_starts += 1;
    this.starts += 1;
  }
}

export class FollowupQuestionGenCallbackHandler extends BaseCallbackHandler {
  chain_starts: number = 0;
  chain_ends: number = 0;
  llm_starts: number = 0;
  llm_ends: number = 0;
  llm_streams: number = 0;
  tool_starts: number = 0;
  tool_ends: number = 0;
  agent_ends: number = 0;
  starts: number = 0;
  ends: number = 0;
  errors: number = 0;

  constructor(private websocket: any) {
    super();
  }

  async on_llm_new_token(token: string, kwargs: any): Promise<void> {
    const resp: ChatResponse = {
      sender: 'bot',
      message: token,
      type: 'stream_followup'
    };
    console.log(token);
    await this.websocket.send_json(resp);
  }

  on_llm_start(serialized: any, prompts: string[], kwargs: any): void {
    this.llm_starts += 1;
    this.starts += 1;
  }

  on_llm_end(response: LLMResult, kwargs: any): void {
    this.llm_ends += 1;
    this.ends += 1;
  }

  on_llm_error(error: Exception | KeyboardInterrupt, kwargs: any): void {
    this.errors += 1;
  }

  on_chain_start(serialized: any, inputs: any, kwargs: any): void {
    this.chain_starts += 1;
    this.starts += 1;
  }

  on_chain_end(outputs: any, kwargs: any): void {
    this.chain_ends += 1;
    this.ends += 1;
  }

  on_chain_error(error: Exception | KeyboardInterrupt, kwargs: any): void {
    this.errors += 1;
  }

  on_tool_start(serialized: any, input_str: string, kwargs: any): void {
    this.tool_starts += 1;
    this.starts += 1;
  }

  on_tool_end(output: string, kwargs: any): void {
    this.tool_ends += 1;
    this.ends += 1;
  }

  on_tool_error(error: Exception | KeyboardInterrupt, kwargs: any): void {
    this.errors += 1;
  }

  on_text(text: string, kwargs: any): void {
    this.text += 1;
  }

  on_agent_finish(finish: AgentFinish, kwargs: any): void {
    this.agent_ends += 1;
    this.ends += 1;
  }

  on_agent_action(action: AgentAction, kwargs: any): any {
    this.tool_starts += 1;
    this.starts += 1;
  }
}